### DISCUSSION TITLE
Incorporating Bot into Flask App

### DISCUSSION THREAD

**Discussion by shapiroj18** [python-telegram-bot/python-telegram-bot#2350] [Q&A]
I'm hoping to build out a bot within a flask application. Per the suggestion of @Bibo-Joshi, I followed the [Custom Solution skeleton ](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Webhooks), but am falling a bit short.

My code can be seen here in [pastebin](https://pastebin.com/vw88K9uk). Using v13.1.

**Comment by Bibo-Joshi** [comment]
Hi. As discussed in the user group `update = Update.de_json(request.get_json(force=True), bot)` should lead to `update.effective_message.bot` being the same object as the `bot` passed to `de_json`.  Could you verify that manually after forming the update, i.e. check if the update has an `effective_message` and if yeas, print out `update.effective_message.bot is bot`? if that is `True`, please print double check, what `bot` is. If it's not `True`, please try check, what `update.effective_message.bot` is. The function probably has a name, maybe that helps …

**Reply by shapiroj18** [reply]
Thanks for the suggestion @Bibo-Joshi . I updated the flask route to include the following for debugging purposes:

```
@app.route(f"/{auth_key}", methods=["POST"])
def webhook_telegram():
    update = Update.de_json(request.get_json(force=True), bot)
    update_queue.put(update)
    print(f"update:\n{update}")
    print(f"update.effective_message:\n{update.effective_message}")
    print(f"update.effective_message.bot:\n{update.effective_message.bot}")
    print(f"bot:\n{bot}")
```

When I hit the webhook, I get the following: 
```
2021-02-01 09:31:24,659 - werkzeug - INFO -  * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
update:
{'update_id': 876700828, 'message': {'message_id': 88, 'date': 1612189893, 'chat': {'id': 1429248182, 'type': 'private', 'first_name': 'Jonathan', 'last_name': 'Shapiro'}, 'text': '/start', 'entities': [{'type': 'bot_command', 'offset': 0, 'length': 6}], 'caption_entities': [], 'photo': [], 'new_chat_members': [], 'new_chat_photo': [], 'delete_chat_photo': False, 'group_chat_created': False, 'supergroup_chat_created': False, 'channel_chat_created': False, 'from': {'id': 1429248182, 'first_name': 'Jonathan', 'is_bot': False, 'last_name': 'Shapiro', 'language_code': 'en'}}}
update.effective_message:
{'message_id': 88, 'date': 1612189893, 'chat': {'id': 1429248182, 'type': 'private', 'first_name': 'Jonathan', 'last_name': 'Shapiro'}, 'text': '/start', 'entities': [{'type': 'bot_command', 'offset': 0, 'length': 6}], 'caption_entities': [], 'photo': [], 'new_chat_members': [], 'new_chat_photo': [], 'delete_chat_photo': False, 'group_chat_created': False, 'supergroup_chat_created': False, 'channel_chat_created': False, 'from': {'id': 1429248182, 'first_name': 'Jonathan', 'is_bot': False, 'last_name': 'Shapiro', 'language_code': 'en'}}
update.effective_message.bot:
<function bot at 0x7fde701c3d40>
bot:
<function bot at 0x7fde701c3d40>
```

**Reply by Bibo-Joshi** [reply]
Allright, the last two prints tell me that `bot` is not what you think it is. please try:

1. printing `bot` directly after line 9 to make sure that it's actually a `Bot` instance there
2. printing `bot.__name__` in line 28. That should give the name of that function. maybe that tells you, what it is

I don't know much about flask, though, so I can't tell, if you e.g. need to pass `bot` to the `webhook_telegram` callback in a different way …

**Reply by shapiroj18** [reply]
For sure -

For 1:
```
bot = Bot(auth_key)
print(f"bot:\n{bot}")
```
Response:
```
bot:
{'id': 1561152901, 'username': 'devlocal_gone_phishing_bot', 'first_name': 'Dev Local Phish Bot'}
```

For 2:
```
@app.route(f"/{auth_key}", methods=["POST"])
def webhook_telegram():
    update = Update.de_json(request.get_json(force=True), bot)
    print(f"bot.__name__:\n{bot.__name__}")
    update_queue.put(update)    
```

Response:
```
bot.__name__:
bot
```

**Reply by Bibo-Joshi** [reply]
Mh, first printout looks good, the second is not too helpful … maybe you need to call `de_json(…, bot())`? Anyway, at this point I'm inclined to say that it's a flask problem, as the issue is to properly access `bot` in `webhook_telegram`.

**Comment by shapiroj18** [comment]
@Bibo-Joshi thanks for all of your help on this - unfortunately there wasn't a good solution here, but from a design perspective I realized it made sense to make the bot application separate from the flask app itself. Now I have it just work with the API I built in Flask.

Appreciate all of this once again!

**Reply by Bibo-Joshi** [reply]
Hey, no. Also sorry, for not replying, must have missed the notification. Regardles of how you proceed:

> For 2:
>
>```python
>@app.route(f"/{auth_key}", methods=["POST"])
>def webhook_telegram():
>    update = Update.de_json(request.get_json(force=True), bot)
>    print(f"bot.__name__:\n{bot.__name__}")
>    update_queue.put(update)    
>```
>Response:
>
>```
>bot.__name__:
>bot
>```

This tells me, that in `webhook_telegram()` the variable `bot` for some reason is a function named `bot`. Don't know what flask does here …
